core/modules/qproc/ChunkSpec.cc

```diff
@@ -57,12 +60,116 @@ std::ostream& operator<<(std::ostream& os, ChunkSpec const& c) {
     return os;
 }
 
+
+/// ChunkSpecVector intersection.
+/// Computes ChunkSpecVector intersection, overwriting dest with the result.
+///
+/// precondition: Elements in ChunkSpecVector should be minimized in the sense
+/// that there should be only one element for a particular chunkId. i.e., for
+/// all ChunkSpec A element, there is no ChunkSpec B element in the same vector
+/// where A.chunkId == B.chunkId.
+void intersectSorted(ChunkSpecVector& dest, ChunkSpecVector const& a) {
+    ChunkSpecVector tmp;
+    tmp.swap(dest);
+    ChunkSpecVector::iterator di = tmp.begin();
+    ChunkSpecVector::iterator de = tmp.end();
+    ChunkSpecVector::const_iterator ai = a.begin();
+    ChunkSpecVector::const_iterator ae = a.end();
+    for(; (di != de) && (ai != ae); ++di) { // march down dest vector
+        // For each item in dest, advance through a to find a matching chunkId.
+        while(ai->chunkId < di->chunkId) {
+            ++ai;
+        }
+
+        if(ai->chunkId == di->chunkId) {
+            ChunkSpec cs = *di;
+            // On a match, perform the intersection.
+            cs.restrict(*ai);
+            if(di->chunkId != ChunkSpec::CHUNKID_INVALID) {
+                dest.push_back(cs);
+            }
+        }
+    }
+}
+
+
+ChunkSpecVector intersect(ChunkSpecVector const& a, ChunkSpecVector const& b) {
+    ChunkSpecVector asort, bsort;
+    asort = a;
+    std::sort(asort.begin(), asort.end());
+    bsort = b;
+    std::sort(bsort.begin(), bsort.end());
+    intersectSorted(asort, bsort);
+    return asort;
+}
+
 ////////////////////////////////////////////////////////////////////////
 // ChunkSpec
 ////////////////////////////////////////////////////////////////////////
 bool ChunkSpec::shouldSplit() const {
     return subChunks.size() > (unsigned)GOOD_SUBCHUNK_COUNT;
 }
+
+ChunkSpec ChunkSpec::intersect(ChunkSpec const& cs) const {
+    ChunkSpec output(*this);
+    output.restrict(cs);
+    return output;
+}
+
+void ChunkSpec::restrict(ChunkSpec const& rhs) {
+    if(chunkId != rhs.chunkId) {
+        throw Bug("ChunkSpec::merge with different chunkId");
+    }
+    std::sort(subChunks.begin(), subChunks.end());
+    Int32Vector rhsSort, output;
+    rhsSort.assign(rhs.subChunks.begin(), rhs.subChunks.end());
+    std::sort(rhsSort.begin(), rhsSort.end());
+    std::set_intersection(
+        subChunks.begin(), subChunks.end(),
+        rhsSort.begin(), rhsSort.end(),
+        std::insert_iterator<Int32Vector>(output, output.end()));
+    subChunks.swap(output);
+}
+
+bool ChunkSpec::operator<(ChunkSpec const& rhs) const {
+    typedef std::vector<int32_t>::const_iterator VecIter;
+
+    if(chunkId < rhs.chunkId) return true;
+    else if(chunkId > rhs.chunkId) return false;
+    else {
+        std::pair<VecIter,VecIter> mism
+            = std::mismatch(subChunks.begin(), subChunks.end(),
+                            rhs.subChunks.begin());
```

Comment:
are subChunks and rhs.subChunks guaranteed to be the same length?  (if rhs.subChunks is shorter, behavior of std::mismatch is I believe undefined?)